### ISSUE TITLE
detect unnecessary copy/missing std::move()

### ISSUE THREAD

**Issue by firewave** [llvm/llvm-project#53489]
Clang-Tidy should warn about the following cases of unnecessary copies:

```cpp
#include <vector>
#include <string>

class A
{
public:
    A(std::vector<std::string> v)
    : v_(std::move(v))
    {}
private:
    std::vector<std::string> v_;
};

int main() {
  std::vector<std::string> v;
  for(int i = 0; i < 10; ++i)
  {
    v.emplace_back("str");
  }
  A a{v}; // missing std::move()
  return 0;
}
```

```cpp
#include <vector>
#include <string>

int main() {
  std::vector<std::string> v;

  std::string s;
  for(int i = 0; i < 10; ++i)
  {
    s += " ";
  }
  v.emplace_back(s); // missing std::move()
  return 0;
}
```

```cpp
#include <vector>

struct A
{
    std::vector<int> v;
};

int main() {
    std::vector<int> v;
    v.emplace_back(1);
    A a;
    a.v = v; // missing std::move()
    return 0;
}
```

This was causing major performance issue on a project I once worked on. It required a lot of profiling and manual code review to find all occurrences so any pointers by a static code analysis tool would helpful.

This warning only makes sense on non-trivially-copyable types and types without side effects.

For objects with side effects it is not safe to judge if the copy was intended or not - even if the object is not used afterwards.

**Comment by Izaron** [comment]
I personally think that this kind of checks is really hard or almost impossible to implement, otherwise it would go straight to the Standard and become mandatory.

> types without side effects

Did you mean "types with copy/move ctors without side effects"? If so, I think we can accept these types, because RVO/NRVO optimizatons don't care if ctors had side effects.

> This was causing major performance issue on a project I once worked on

My approach was to check which types have a lot of objects, and make these types non-copyable by deleting its copy ctors. It completely destroyed every possible source of copying once and for all.

**Comment by firewave** [comment]
> I personally think that this kind of checks is really hard or almost impossible to implement, otherwise it would go straight to the Standard and become mandatory.

A modified version of the first example is being detected by `performance-unnecessary-value-param` if the copied object is a function parameter - so "almost impossible" seems like an exaggeration:
```cpp
#include <string>

class A
{
public:
	A(std::string s)
	: s_(std::move(s))
	{}
private:
	std::string s_;
};

extern void f(std::string s) {
	A a{s}; // performance-unnecessary-value-param
}
```

If I re-add the `std::vector<std::string>` usage it will warn neither for the `emplace_back()` nor for the constructor call.

> Did you mean "types with copy/move ctors without side effects"? If so, I think we can accept these types, because RVO/NRVO optimizatons don't care if ctors had side effects.

Yes. I would also include "dtors without side effects". So basically types which are just used for data storage as seen in the examples.

> My approach was to check which types have a lot of objects, and make these types non-copyable by deleting its copy ctors. It completely destroyed every possible source of copying once and for all.

That would be a way to reduce this but depending on how the type is used that might not be possible. Also that will only be possible with user types.

**Comment by firewave** [comment]
I find the description of the referenced ticket much clearer in terms of what the proposed check should achieve.

And apparently some of the information already seems to be available since clang-tidy checks can determine e.g. if a `std::move()` will actually move - see https://github.com/llvm/llvm-project/issues/57266. So I assume there's information if something was copied as well.

**Comment by Febbe** [comment]
Let's assume I have time and know roughly how to approach this (Have already put work into AST evaluation and `clang-query`). Should I then develop a completely separate check for this, or integrate it into `UnnecessaryValueParamCheck`?

**Comment by JonasToth** [comment]
thats a "definite last use" and theoretically possible to detect and implement a transformation that emits a `std::move` there. If its _wise to do so_ I am not so certain about. Such moves are easy to get overlooked and future refactoring cause garbage memory to be used.

For more broader discussion and thoughts about these things, Herb Sutter has IMHO two good talks:

- [Herb Sutters Talk 1](https://www.youtube.com/watch?v=6lurOCdaj0Y) for a broader discussion about parameters, moving and referencing, forwarding and how its possible to simplify
- [Herb Sutters cppfront](https://www.youtube.com/watch?v=ELeZAKCN4tY) that produces the "move on definite last use" (at 01:26:05)

**Comment by JonasToth** [comment]
@Febbe I believe you need a flow-sensitive analysis, similar to what `bugprone-use-after-move` does (maybe even the same logic!), because you need to find the "definite last use" of a variable.
See the two talks in my previous comment, that explain these things good.

Additionally, there is [Lifetime Paper](https://raw.githubusercontent.com/isocpp/CppCoreGuidelines/master/docs/Lifetime.pdf) that addresses a different issue: How to prevent "use-after-free" and other forms of undefined behaviour guaranteed. This seems like something else first, but logically, its all about following the lifetime of a variable. The paper contains relevant algorithmic ideas about these issues.

The action "emit `std::move`" for the last use is then just a different reaction.
Overall, I still think that such a check/transformation should be done carefully, because its a creator of future undefined behavior do to `use-after-move` issues (but again, there is a check for that too :D)

**Comment by firewave** [comment]
@JonasToth Thanks for the pointers. I remember a former colleague mentioning that Herb Sutter even wrote a whole book just about `std::move` which he had just acquired back then.

If this just catches the most basic cases it should still have quite an positive impact on the code. Similar to my examples - the actual code it was reduced from wasn't much more complex.

I also like the "no use after copy" terminology by @Febbe in the other ticket which sums up the basic case quite clearly. 🙂

**Comment by firewave** [comment]
> Yes, but espexially because `std::move` is such a dragon, a check suggesting it should be rock solid. Emitting warnings/fixits that suggest a `std::move` somewhere must not have false positives.

As mentioned in https://github.com/llvm/llvm-project/issues/53489#issuecomment-1034376991 some of the suggested warnings basically already exists. It just doesn't fire on local variables.

```cpp
#include <string>

class A
{
public:
	A(std::string s)
	: s_(std::move(s))
	{}
private:
	std::string s_;
};

extern void f1(std::string s) {
	A a{s}; // performance-unnecessary-value-param
}

extern void f2()
{
        std::string s;
        A a{s}; // no warning
}
```

Interestingly if the parameter is used it is not detected correctly. Or it is a known limitation as it tries to be "rock solid".

```cpp
extern void f3(std::string s) {
    if (!s.empty()) // performance-unnecessary-value-param
    {
	    A a{s}; // no warning
    }
}
```

I filed #57908 about that.

https://godbolt.org/z/5hG31nos7

Sorry if I repeated myself too much from above.

**Comment by Febbe** [comment]
@JonasToth also thank you for those references. They are quite interesting, but this is something which might never make it into the standard :/.

Maybe an idea for another check, which can evaluate Funktions with some defines: `#define in` `#define out` `#define inout` ... and do the checks, suggested by Herb Sutter. 

> @Febbe I believe you need a flow-sensitive analysis, similar to what bugprone-use-after-move does (maybe even the same logic!), because you need to find the "definite last use" of a variable.

I already searched for other checks like bugprone-use-after-move, some of them used the position in the source code, which is not that stable I think. But for prototyping, it might be ok. For example, a definite last use in an if else can't be detected by this.

Also, I found another case where this check should be relevant: Not only a "not use/delete after copy", but also any "assignment after copy" should trigger it.

> Yes, but especially because std::move is such a dragon, a check suggesting it should be rock solid. Emitting warnings/fixits that suggest a std::move somewhere must not have false positives. 

Warning behind an optional feature are ok. Then a programmer can decide by himself it this warning is sane.

**Comment by JonasToth** [comment]
> @JonasToth also thank you for those references. They are quite interesting, but this is something which might never make it into the standard :/.

I am not so sure about that. All the copy elision and guaranteed RVO things in the standard are kinda similar. And returning by value has a guaranteed `move` if the other things dont trigger. Returning from a function is very similar to a definite last use. If the function would be inlined, it would even look the same in terms of C++ code (if you would manually inline the function in your code by moving the code block).

But I don't know if there is work on it and if so, what status/consensus level it has.

For everything custom (your macro example): you can try this out as prototype or so and maybe discuss it in the `cppcoreguidelines` (or so). But I think `clang-tidy` should base such a check on some coding standard, and in the best case on the `cppcoreguidelines`. Afterall, this is a personal thing of Herb Sutter and its not clear in which direction the whole community goes.

> For example, a definite last use in an if else can't be detected by this

It should be detectable. You are building a graph, which each usage being a node and control flow splits creating branches (control-flow-graph, kind of a tree).
All leave nodes are "definite last use" and those usages can `move`.

**Comment by Febbe** [comment]
> It should be detectable. You are building a graph, which each usage being a node and control flow splits creating branches (control-flow-graph, a tree if you have only statements and if/else). Creating the CFG is already implemented, please don't reinvent :)
All leave nodes are "definite last use" and those usages can move.
From those leave nodes, return and throw (maybe?!) should be ignored.

Ok I have a first approach done, CFG works, basic leaves are detected, but I found some cases, where a move will introduce bugs. For example, when moving a shared_ptr containing a guard. So we must prohibit all (automatic) moves (fixes) for guards.

Also, detecting a copy before an assignment does not work yet. 

For information, the current branch is: https://github.com/Febbe/llvm-project/tree/test-perf-un-copy-before-delete
But it can take some days, until I have time again.

**Comment by firewave** [comment]
The branch is quite out-of-date. I think it would make sense to rebase it before somebody starts messing around with it.

**Comment by Febbe** [comment]
I was also confused, that I was 20k commits behind. Rebased it, and also add the warning, before it just printed the case on the console.

**Comment by firewave** [comment]
It does not compile for me. I am using clang-15.
```
/mnt/s/GitHub/llvm-project/clang-tools-extra/clang-tidy/performance/UnnecessaryCopyOnLastUseCheck.cpp:75:2: warning: extra ';' outside of a function is incompatible with C++98 [-Wc++98-compat-extra-semi]
};
 ^
/mnt/s/GitHub/llvm-project/clang-tools-extra/clang-tidy/performance/UnnecessaryCopyOnLastUseCheck.cpp:186:7: error: no matching constructor for initialization of 'utils::IncludeInserter'
      Inserter(Options.getLocalOrGlobal("IncludeStyle",
      ^        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/mnt/s/GitHub/llvm-project/clang-tools-extra/clang-tidy/performance/../utils/IncludeInserter.h:55:7: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'std::enable_if_t<std::is_enum<IncludeStyle>::value, IncludeStyle>' (aka 'clang::tidy::utils::IncludeSorter::IncludeStyle') to 'const clang::tidy::utils::IncludeInserter' for 1st argument
class IncludeInserter {
      ^
/mnt/s/GitHub/llvm-project/clang-tools-extra/clang-tidy/performance/../utils/IncludeInserter.h:55:7: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'std::enable_if_t<std::is_enum<IncludeStyle>::value, IncludeStyle>' (aka 'clang::tidy::utils::IncludeSorter::IncludeStyle') to 'clang::tidy::utils::IncludeInserter' for 1st argument
/mnt/s/GitHub/llvm-project/clang-tools-extra/clang-tidy/performance/../utils/IncludeInserter.h:62:12: note: candidate constructor not viable: requires 2 arguments, but 1 was provided
  explicit IncludeInserter(IncludeSorter::IncludeStyle Style,
           ^
/mnt/s/GitHub/llvm-project/clang-tools-extra/clang-tidy/performance/UnnecessaryCopyOnLastUseCheck.cpp:188:7: error: no matching constructor for initialization of 'const std::vector<std::string>' (aka 'const vector<basic_string<char>>')
      AllowedTypes(
      ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_vector.h:494:7: note: candidate constructor not viable: no known conversion from 'std::vector<StringRef>' to 'const std::vector<std::basic_string<char>>::allocator_type' (aka 'const std::allocator<std::basic_string<char>>') for 1st argument
      vector(const allocator_type& __a) _GLIBCXX_NOEXCEPT
      ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_vector.h:507:7: note: candidate constructor not viable: no known conversion from 'std::vector<StringRef>' to 'std::vector::size_type' (aka 'unsigned long') for 1st argument
      vector(size_type __n, const allocator_type& __a = allocator_type())
      ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_vector.h:550:7: note: candidate constructor not viable: no known conversion from 'vector<llvm::StringRef, allocator<llvm::StringRef>>' to 'const vector<std::basic_string<char>, allocator<std::basic_string<char>>>' for 1st argument
      vector(const vector& __x)
      ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_vector.h:569:7: note: candidate constructor not viable: no known conversion from 'vector<llvm::StringRef, allocator<llvm::StringRef>>' to 'vector<std::basic_string<char>, allocator<std::basic_string<char>>>' for 1st argument
      vector(vector&&) noexcept = default;
      ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_vector.h:622:7: note: candidate constructor not viable: no known conversion from 'std::vector<StringRef>' to 'initializer_list<std::vector<std::basic_string<char>, std::allocator<std::basic_string<char>>>::value_type>' (aka 'initializer_list<std::basic_string<char>>') for 1st argument
      vector(initializer_list<value_type> __l,
      ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_vector.h:484:7: note: candidate constructor not viable: requires 0 arguments, but 1 was provided
      vector() = default;
      ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_vector.h:572:7: note: candidate constructor not viable: requires 2 arguments, but 1 was provided
      vector(const vector& __x, const allocator_type& __a)
      ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_vector.h:604:7: note: candidate constructor not viable: requires 2 arguments, but 1 was provided
      vector(vector&& __rv, const allocator_type& __m)
      ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_vector.h:519:7: note: candidate constructor not viable: requires at least 2 arguments, but 1 was provided
      vector(size_type __n, const value_type& __value,
      ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_vector.h:582:7: note: candidate constructor not viable: requires 3 arguments, but 1 was provided
      vector(vector&& __rv, const allocator_type& __m, true_type) noexcept
      ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_vector.h:586:7: note: candidate constructor not viable: requires 3 arguments, but 1 was provided
      vector(vector&& __rv, const allocator_type& __m, false_type)
      ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_vector.h:650:2: note: candidate constructor template not viable: requires at least 2 arguments, but 1 was provided
        vector(_InputIterator __first, _InputIterator __last,
        ^
/mnt/s/GitHub/llvm-project/clang-tools-extra/clang-tidy/performance/UnnecessaryCopyOnLastUseCheck.cpp:303:53: error: no viable conversion from 'const std::vector<std::string>' (aka 'const vector<basic_string<char>>') to 'ArrayRef<llvm::StringRef>'
                utils::options::serializeStringList(AllowedTypes));
                                                    ^~~~~~~~~~~~
/mnt/s/GitHub/llvm-project/llvm/include/llvm/ADT/ArrayRef.h:42:40: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'const std::vector<std::string>' (aka 'const vector<basic_string<char>>') to 'const llvm::ArrayRef<llvm::StringRef> &' for 1st argument
  class LLVM_GSL_POINTER [[nodiscard]] ArrayRef {
                                       ^
/mnt/s/GitHub/llvm-project/llvm/include/llvm/ADT/ArrayRef.h:42:40: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'const std::vector<std::string>' (aka 'const vector<basic_string<char>>') to 'llvm::ArrayRef<llvm::StringRef> &&' for 1st argument
/mnt/s/GitHub/llvm-project/llvm/include/llvm/ADT/ArrayRef.h:71:18: note: candidate constructor not viable: no known conversion from 'const std::vector<std::string>' (aka 'const vector<basic_string<char>>') to 'llvm::NoneType' for 1st argument
    /*implicit*/ ArrayRef(NoneType) {}
                 ^
/mnt/s/GitHub/llvm-project/llvm/include/llvm/ADT/ArrayRef.h:74:18: note: candidate constructor not viable: no known conversion from 'const std::vector<std::string>' (aka 'const vector<basic_string<char>>') to 'const llvm::StringRef &' for 1st argument
    /*implicit*/ ArrayRef(const T &OneElt)
                 ^
/mnt/s/GitHub/llvm-project/llvm/include/llvm/ADT/ArrayRef.h:115:18: note: candidate constructor not viable: no known conversion from 'const std::vector<std::string>' (aka 'const vector<basic_string<char>>') to 'const std::initializer_list<StringRef> &' for 1st argument
    /*implicit*/ ArrayRef(const std::initializer_list<T> &Vec)
                 ^
/mnt/s/GitHub/llvm-project/llvm/include/llvm/ADT/ArrayRef.h:89:18: note: candidate template ignored: could not match 'SmallVectorTemplateCommon' against 'vector'
    /*implicit*/ ArrayRef(const SmallVectorTemplateCommon<T, U> &Vec)
                 ^
/mnt/s/GitHub/llvm-project/llvm/include/llvm/ADT/ArrayRef.h:95:18: note: candidate template ignored: could not match 'llvm::StringRef' against 'std::basic_string<char>'
    /*implicit*/ ArrayRef(const std::vector<T, A> &Vec)
                 ^
/mnt/s/GitHub/llvm-project/llvm/include/llvm/ADT/ArrayRef.h:100:28: note: candidate template ignored: could not match 'array' against 'vector'
    /*implicit*/ constexpr ArrayRef(const std::array<T, N> &Arr)
                           ^
/mnt/s/GitHub/llvm-project/llvm/include/llvm/ADT/ArrayRef.h:105:28: note: candidate template ignored: could not match 'const llvm::StringRef[N]' against 'const std::vector<std::string>' (aka 'const vector<basic_string<char>>')
    /*implicit*/ constexpr ArrayRef(const T (&Arr)[N]) : Data(Arr), Length(N) {}
                           ^
/mnt/s/GitHub/llvm-project/llvm/include/llvm/ADT/ArrayRef.h:125:5: note: candidate template ignored: could not match 'ArrayRef' against 'vector'
    ArrayRef(const ArrayRef<U *> &A,
    ^
/mnt/s/GitHub/llvm-project/llvm/include/llvm/ADT/ArrayRef.h:134:18: note: candidate template ignored: could not match 'SmallVectorTemplateCommon' against 'vector'
    /*implicit*/ ArrayRef(
                 ^
/mnt/s/GitHub/llvm-project/llvm/include/llvm/ADT/ArrayRef.h:143:5: note: candidate template ignored: could not match 'type-parameter-0-0 *' against 'std::basic_string<char>'
    ArrayRef(const std::vector<U *, A> &Vec,
    ^
/mnt/s/GitHub/llvm-project/clang-tools-extra/clang-tidy/performance/../utils/OptionsUtils.h:28:53: note: passing argument to parameter 'Strings' here
std::string serializeStringList(ArrayRef<StringRef> Strings);
                                                    ^
```

**Comment by firewave** [comment]
That's understandable. Thanks a lot.

On the first quick run I see a couple of false positives but also a lot of true positives which are easily to miss if you just do a code review. Also some awkward cases which makes it obvious which such a check is complex.

I think I also see some false negatives in other checks.

So even if this might not make it upstream I definitely see value in maintaining the branch.

**Comment by Febbe** [comment]
> On the first quick run I see a couple of false positives but also a lot of true positives which are easily to miss if you just do a code review.

That's awesome, did you filter out my debug output, which prints literally every copy?

> Also, some awkward cases which makes it obvious which such a check is complex.

Do you have some examples, or as alternative the lines of the occurrences in some open source projects?
Including such cases in the tests would decrease false positives. 

> So even if this might not make it upstream I definitely see value in maintaining the branch.

There could be some concerns about automatic fix ups. The reason no one did this is, that it is extremely hard to say, if an extra move will first increase performance, second, does not change program behavior. Therefore, honestly every possible last use must be checked by a programmer.   
But when I find a way to disable automatic fix ups for all or at least uncertain cases, I don't see a reason to block this.

Btw. my next steps are: 
 - Add a function name list property, for false positive functions, like `false_positive_function(AnyType must_copy_value)`.
 - Add a DeclStmt list property for types, which must not be moved, like: `struct my_special_RAII_Guard`
 - Property to enable warnings for all cases, even for unlikely ones.
 - Add more Test cases.

**Comment by firewave** [comment]
> That's awesome, did you filter out my debug output, which prints literally every copy?

Obviously ;-)

> Do you have some examples, or as alternative the lines of the occurrences in some open source projects?
> Including such cases in the tests would decrease false positives.

I will provide those later. Will take a while though.

The main false positive was variables that were being returned. They were non-`const` and should be implicitly moved and not be copies at all. In several case the type of the variable was `std::string`.

FYI I scanned [Cppcheck](https://github.com/danmar/cppcheck) with it. In case you want to take a look yourself.

**Comment by Febbe** [comment]
I made some fixes to decrease false positives.
Detections on return are weird, the matcher searches for copy constructors, implicit or not. Didn't had some when I scanned llvm itself😅. But it could be, that there is an std::forward missing in an template, or that the std::string is a rvalue-reference type. Might be an additional improvement.

**Comment by firewave** [comment]
Cppcheck uses C++11 as configured standard whereas the LLVM code base uses C++17 - maybe that makes a difference.

**Comment by firewave** [comment]
Here's a branch with the first round of fixes I applied. I have not tested them beside compilation yet. It is also not based on your lasted code yet.

**Comment by firewave** [comment]
I did not fix warnings where an object is copied multiple times beforehand and if a copy might be introduced in the future:

```cpp

void cb1(std::string);
void cb2(std::string);
void cb3(std::string);

void f()
{
std::string s;
cb1(s);
cb2(s);
cb3(s); // warning
}
```

Maybe this could be made configurable that it is only reported if there is a single copy or with a threshold of the previous amount of copies.

**Comment by Febbe** [comment]
> Maybe this could be made configurable that it is only reported if there is a single copy or with a threshold of the previous amount of copies.

I am not sure, if that's desirable. First, there is also a check for moves before uses and a programmer who adds a call hopefully moves the `std::move`.
But either way, this option should be added, after the check is accepted, I think.

> It is also not based on your lasted code yet. 

It may be worth it, because I removed some cases, where a move is obsolete (e.g. move on trivially copyable type is the same as a move). Also, elidable copies on a return are not flagged anymore. But missing moves on a return from rvalue-references are not generated yet.

**Comment by firewave** [comment]
> I am not sure, if that's desirable. First, there is also a check for moves before uses and a programmer who adds a call hopefully moves the `std::move`.

In most cases it was code with lots of calls and only part of them used the parameter and new calls might be added in the future so you would need to move the `std::move()` around. It is also very cold code. `NOLINTBEGIN/NOLINTEND` could be used but it seems like a common pattern so the annotation could be avoided.

> It may be worth it,

I still had to compile it. And it doesn't :(

```
 In file included from /mnt/s/GitHub/llvm-project/clang-tools-extra/clang-tidy/performance/PerformanceTidyModule.cpp:9:
In file included from /mnt/s/GitHub/llvm-project/clang-tools-extra/clang-tidy/performance/../ClangTidy.h:12:
In file included from /mnt/s/GitHub/llvm-project/clang-tools-extra/clang-tidy/performance/../ClangTidyDiagnosticConsumer.h:12:
In file included from /mnt/s/GitHub/llvm-project/clang-tools-extra/clang-tidy/performance/../ClangTidyOptions.h:12:
In file included from /mnt/s/GitHub/llvm-project/llvm/include/llvm/ADT/IntrusiveRefCntPtr.h:66:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/memory:80:
/usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/unique_ptr.h:79:16: error: invalid application of 'sizeof' to an incomplete type 'clang::CFG'
        static_assert(sizeof(_Tp)>0,
                      ^~~~~~~~~~~
/usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/unique_ptr.h:292:4: note: in instantiation of member function 'std::default_delete<clang::CFG>::operator()' requested here
          get_deleter()(std::move(__ptr));
          ^
/mnt/s/GitHub/llvm-project/llvm/include/llvm/ADT/DenseMap.h:368:25: note: in instantiation of member function 'std::unique_ptr<clang::CFG>::~unique_ptr' requested here
        P->getSecond().~ValueT();
                        ^
/mnt/s/GitHub/llvm-project/llvm/include/llvm/ADT/DenseMap.h:754:11: note: in instantiation of member function 'llvm::DenseMapBase<llvm::DenseMap<const clang::FunctionDecl *, std::unique_ptr<clang::CFG>>, const clang::FunctionDecl *, std::unique_ptr<clang::CFG>, llvm::DenseMapInfo<const clang::FunctionDecl *>, llvm::detail::DenseMapPair<const clang::FunctionDecl *, std::unique_ptr<clang::CFG>>>::destroyAll' requested here
    this->destroyAll();
          ^
/mnt/s/GitHub/llvm-project/clang-tools-extra/clang-tidy/performance/UnnecessaryCopyOnLastUseCheck.h:29:7: note: in instantiation of member function 'llvm::DenseMap<const clang::FunctionDecl *, std::unique_ptr<clang::CFG>>::~DenseMap' requested here
class UnnecessaryCopyOnLastUseCheck : public ClangTidyCheck {
      ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/unique_ptr.h:292:4: note: in instantiation of member function 'std::default_delete<clang::tidy::performance::UnnecessaryCopyOnLastUseCheck>::operator()' requested here
          get_deleter()(std::move(__ptr));
          ^
/mnt/s/GitHub/llvm-project/clang-tools-extra/clang-tidy/performance/../ClangTidyModule.h:62:35: note: in instantiation of member function 'std::unique_ptr<clang::tidy::performance::UnnecessaryCopyOnLastUseCheck>::~unique_ptr' requested here
                           return std::make_unique<CheckType>(Name, Context);
                                  ^
/mnt/s/GitHub/llvm-project/clang-tools-extra/clang-tidy/performance/PerformanceTidyModule.cpp:61:20: note: in instantiation of function template specialization 'clang::tidy::ClangTidyCheckFactories::registerCheck<clang::tidy::performance::UnnecessaryCopyOnLastUseCheck>' requested here
    CheckFactories.registerCheck<UnnecessaryCopyOnLastUseCheck>(
                   ^
/mnt/s/GitHub/llvm-project/clang-tools-extra/clang-tidy/performance/UnnecessaryCopyOnLastUseCheck.h:20:7: note: forward declaration of 'clang::CFG'
class CFG;
      ^
```

I am using Clang 15 to compile.

**Comment by Febbe** [comment]
I rebased it again, so I can provide a patch via arcanist. MAy be that they changed again the ABI.

**Comment by Febbe** [comment]
So, I have fixed it. 
Somehow, clang has a problem with `unique_ptr`s of incomplete type in containers like llvm::DenseMap.  Don't know if this is a clang or msvc bug.

I also created a diff on [phabricator](https://reviews.llvm.org/D137205)

**Comment by firewave** [comment]
Thanks. I have not taken a look yet. I got side-tracked. Hopefully will take the next look on the weekend.

As I have fixed lots of such issues already (based on code reviews) I want to run it on an older version of the code base to check for false negatives.

**Comment by Febbe** [comment]
> Somehow, clang has a problem with `unique_ptr`s of incomplete type in containers like llvm::DenseMap. Don't know if this is a clang or msvc bug.

Was 100% sure I had a destructor, so I even checked that :D. Seems like MSVC does not diagnose that.

**Comment by mclow** [comment]
> > Somehow, clang has a problem with `unique_ptr`s of incomplete type in containers like llvm::DenseMap. Don't know if this is a clang or msvc bug.

What you mean is that "libstdc++, when compiled with clang" has a problem...
But I suspect that libc++ would do the same thing.

And I wouldn't call it a problem, but rather a requirement, that when you instantiate `std::default_delete<T[]>::operator(U *)`, `U` must be complete. See https://wg21.link/unique.ptr.dltr.dflt1, paragraph 3

**Comment by Febbe** [comment]
> > > Somehow, clang has a problem with `unique_ptr`s of incomplete type in containers like llvm::DenseMap. Don't know if this is a clang or msvc bug.
> 
> What you mean is that "libstdc++, when compiled with clang" has a problem... But I suspect that libc++ would do the same thing.
> 
> And I wouldn't call it a problem, but rather a requirement, that when you instantiate `std::default_delete<T[]>::operator(U *)`, `U` must be complete. See https://wg21.link/unique.ptr.dltr.dflt1, paragraph 3

Yes, I know that, I expected, that the msvc's stl would do such a check also. Or I had actually a destructor and the problem was, that I explicitly default constructed the map. But I can't tell now, since I rebased my commits.

**Comment by firewave** [comment]
Still a bit side-tracked and also out for a few days. Might take another week until I will have another look. Sorry about that.

**Comment by firewave** [comment]
How can I find the branch I need to checkout for a Phabricator differential? I did not find anything on the page even after registering. There also appears to be nothing in the user guide.

**Comment by EugeneZelenko** [comment]
@firewave: You could download patch from Differential and apply it to you copy of LLVM code.

**Comment by firewave** [comment]
@EugeneZelenko Thanks.

I built the latest version and updated my branch with the fixes. Also posted some false positives and a crash on Phabricator.

On the false negative side I have identified the commits with the earlier manual fixes I did but will probably take a few days to got through them.

Also of the initial cases I posted only the first one is being detected so far.

**Comment by firewave** [comment]
Some more false negatives:

```cpp
#include <set>
#include <string>

int main() {
	std::string s;

	std::set<std::string> v;
	v.insert(s); // missing std::move()
	return 0;
}
```

```cpp
#include <string>

class C
{
public:
    void f(std::string s) {
        if (!s_p)
            s_p = new std::string(s); // performance-unnecessary-copy-on-last-use
        else
            *s_p = s; // missing std::move()
    }
private:
    std::string* s_p;
};

int main() {
    C c;
    c.f({});
    return 0;
}
```

```cpp
#include <list>
#include <string>

struct S
{
    std::list<std::string> l_s;
};

int main() {

    for (int i = 0; i < 2; i++) // disable loop for warning to appear
    {
        std::string str;
        S s;
        s.l_s = {str}; // missing std::move()
    }
    return 0;
}
```

**Comment by firewave** [comment]
Another false negatives:

```cpp
#include <string>

std::string f(std::string);

int main() {
    std::string s;
    s = f(s); // performance-unnecessary-copy-on-last-use
    s = f(s); // missing std::move()
    return s.size();
}
```

**Comment by firewave** [comment]
More false negatives:

```cpp
#include <array>
#include <string>

void f1()
{
    std::array<std::string, 2> arr;
    std::string s1 = arr[0]; // missing std::move()
    std::string s2 = arr[1]; // missing std::move()
}
```

```cpp
#include <string>

void f1()
{
    std::string arr[2];
    std::string s1 = arr[0]; // missing std::move()
    std::string s2 = arr[1]; // missing std::move()
}
```

**Comment by firewave** [comment]
FYI @PiotrZSL put up an old implementation of such a check at https://reviews.llvm.org/D137205#4225678.
